src/main/scala/xiangshan/frontend/icache/ICache.scala

```diff
@@ -574,9 +574,10 @@ class ICacheImp(outer: ICache) extends LazyModuleImp(outer) with HasICacheParame
   private val ftqPrefetch = WireInit(0.U.asTypeOf(new IPrefetchReq))
   ftqPrefetch.fromFtqICacheInfo(io.ftqPrefetch.req.bits)
   // software prefetch has higher priority
-  prefetcher.io.req.valid  := softPrefetchValid || io.ftqPrefetch.req.valid
-  prefetcher.io.req.bits   := Mux(softPrefetchValid, softPrefetch, ftqPrefetch)
-  io.ftqPrefetch.req.ready := prefetcher.io.req.ready && !softPrefetchValid
+  prefetcher.io.req.valid            := softPrefetchValid || io.ftqPrefetch.req.valid
+  prefetcher.io.req.bits             := Mux(softPrefetchValid, softPrefetch, ftqPrefetch)
+  io.ftqPrefetch.req.ready           := prefetcher.io.req.ready && !softPrefetchValid
+  prefetcher.io.exceptionFromBackend := ExceptionType.fromFtq(io.fetch.req.bits)
```

Comment:
I do not think this is correct:
https://github.com/OpenXiangShan/XiangShan/blob/a6da536392f0f3c14ca05ec883f517c65bd48e87/src/main/scala/xiangshan/frontend/NewFtq.scala#L902-L904

`io.fetch.req.bits` is selected by `ifuPtr`, but prefetchPipe should use request selected by `pfPtr`.

And, since backend exceptions are sent to prefetchPipe s0 now, it will be sent to mainPipe through WayLookup, together with metadata & iTLB results. There should be no need to send it to mainPipe again:
https://github.com/OpenXiangShan/XiangShan/blob/a6da536392f0f3c14ca05ec883f517c65bd48e87/src/main/scala/xiangshan/frontend/icache/ICacheMainPipe.scala#L170

Reply:
ifuPtr and pfPtr should be the same when redirecting, so this should indeed cancel prefetch when there is exception, but it might cancel more prefetch than expected. Anyway, I'm planning a complete refactor from the start point in FTQ and fix this problem in a safer way.